### ISSUE TITLE
Jump in prime counting

### ISSUE THREAD

**Issue by srdeaton** [jwood000/RcppAlgos#14]
Hi, 
Was working with your implementation, doing some detailed calculations. Thought had bug in my code, but tracked it to the prime counting function; specifically to region [1*10^9,1.1*10^9]. The start is hard to find, has somewhat gradual increase, but the end is not, a sharper fall. The end jump is in [1.072*10^9,1.074*10^9] with an error of about 12,400.
Attached some code showing this, and a plot of my function's output which alerted me. For the paper I'm working on, plan to show output for entire computable range; I can skip over this region, but would look strange in plots w/o explanation. Have not decide how to proceed; skip & explain or show & explain. Either way, it's still a good implementation. Don't think there a too many such regions, < 10^16, especially that far off. At least from what I could tell so far.
If you can fix it within the next month, at most, then I can redo calculations in my paper before submission.
Thanks,
Shaun
![PrimeTruncation_Error](https://user-images.githubusercontent.com/64863347/81211886-a937dc00-8f99-11ea-87cf-dc42f21e8b93.png)
[PrimeCounting_ErrorDeviations.txt](https://github.com/jwood000/RcppAlgos/files/4588590/PrimeCounting_ErrorDeviations.txt)

**Comment by srdeaton** [comment]
Note: In the plot above, it not the prime count function, but another I was working on that altered   me to the issue. The following is a plot of the prime counting function itself, for the end of the region shown above.
![PrimeCount_Error_EndOfRegion](https://user-images.githubusercontent.com/64863347/81216370-6af1eb00-8fa0-11ea-8f24-35213023fd7b.png)

**Comment by jwood000** [comment]
Hi @srdeaton,

Thanks for reporting this. Could you provide your `sessionInfo()` (Just type `sessionInfo()` in the console)?

Thanks again,
Joseph

**Comment by srdeaton** [comment]
Hi Joseph,

Here is the session info you asked for:

R version 3.5.1 (2018-07-02)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United
States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United
States.1252

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] RcppAlgos_2.0.3      RevoUtils_11.0.1     RevoUtilsMath_11.0.0

loaded via a namespace (and not attached):
[1] compiler_3.5.1 tools_3.5.1    yaml_2.2.0     Rcpp_0.12.18



I used three different machines, and got same result. This is the PC I use
at work. At home also using a newer, but not newest version of R.

thanks,
Shaun

P.S. Was wondering. Is there a version of the prime count function for
vector inputs, containing increasing integers, such that the function will
not restart its count for each integer?
       Would be good for checking a lot of larger values in a shorter time.
Was going to use prime sieve function & just make my own; but the runtime
of the current implementation
       worked well enough. Additionally, my background is in math &
physics, with programming skills to construct algorithms in these areas.
But, weakness is in anything beyond core
      algorithm building, bad or not nearly as knowledgeable at what you,
and others, would refer to as programming. So, thought my version might not
be up to snuff anyway, at least not
      without a lot of work.
      Thanks again.

**Comment by jwood000** [comment]
@srdeaton,

### TL;DR

This was fixed in version `2.3.0`. N.B. the current version is `2.4.1`. You might want to update [R](<https://cran.r-project.org/bin/windows/base/>) as well. The current version is `4.0.0 (Arbor Day)`.

### The Nitty Gritty

Your problem is that you are using a very old version (almost 2 years old). I recall running into the very same problem while I was doing a major overhaul to `primeCount` and `primeSieve`. The code change was pretty drastic from version `2.2.0` to `2.3.0`. I scoured the commits to see if there were any notes like "Fixed prime count issue around a billion", but was unsuccessful.  I should have opened up an issue when I found it (that's on me). Since then, I've tried to be better about documenting bugs fixes and documentation in general.

I can confirm your results. Below is an example using version `2.2.0`:

```r
packageurl <- "http://cran.r-project.org/src/contrib/Archive/RcppAlgos/RcppAlgos_2.2.0.tar.gz"
install.packages(packageurl, repos=NULL, type="source")

library(RcppAlgos)
M1 <- 0.9 * 10^9
M2 <- 1.1 * 10^9
n <- round(seq(from=M1,to=M2,by=100000))
length(n)
# [1] 2001
plot(n,primeCount_Eval(n),type='l')
```

![algos2 2 0](https://user-images.githubusercontent.com/8388370/81355094-55152080-909b-11ea-9d89-abdae07eaafe.png)

Here is an example using version `2.3.0` (I first restarted my `R` session):

```r
packageurl <- "http://cran.r-project.org/src/contrib/Archive/RcppAlgos/RcppAlgos_2.3.0.tar.gz"
install.packages(packageurl, repos=NULL, type="source")

library(RcppAlgos)
M1 <- 1.073*10^9
M2 <- 1.0735*10^9
n <- round(seq(from=M1,to=M2,by=200))
length(n)
# [1] 2500
plot(n,primeCount_Eval(n),type='l')
```

![algos2 3 0](https://user-images.githubusercontent.com/8388370/81355193-afae7c80-909b-11ea-8fcc-f35629e6ff03.png)

### Validate Using Kim Walisch's primecount library

First, we generate some random data:

```r
set.seed(123)
a <- sample(M1:M2, 10000)
write.csv(sort(a), file = "testWithWalisch.csv", row.names = F)
```

Then, on the command line, we run the Kim Walisch `primecount` on all of the values generated above:

```bash
touch primecount_walisch_output.txt
while IFS=, read -r col1; do primecount-4/primecount "$col1" >> primecount_walisch_output.txt; done < RcppAlgos/testWithWalisch.csv
```

Finally, we read the results into `R` and validate `RcppAlgos::primeCount` on the same input:

```r
fromWalisch <- read.table("../primecount_walisch_output.txt")
dim(fromWalisch)
# [1] 10000     1
head(fromWalisch)
#         V1
# 1 54364435
# 2 54364436
# 3 54364441
# 4 54364444
# 5 54364445
# 6 54364447
canon <- fromWalisch$V1
head(canon)
# [1] 54364435 54364436 54364441 54364444 54364445 54364447
algosImpl <- sapply(sort(a), primeCount)
head(algosImpl)
# [1] 54364435 54364436 54364441 54364444 54364445 54364447

identical(canon, algosImpl)
# [1] TRUE
```

**Comment by srdeaton** [comment]
I should have updated everything beforehand, on me too.
Much appreciated, will have to give you a shout-out in the thank-you
section of my paper.

thanks,
Shaun

**Comment by jwood000** [comment]
@srdeaton,

The shout-out is much appreciated, however I would be remiss if I did not mention @kimwalisch. Quite simply, the prime number capabilities in `RcppAlgos` would not be possible without his excellent work.

Also, to address your question: _"Is there a version of the prime count function for
vector inputs, containing increasing integers, such that the function will
not restart its count for each integer?"_

First off, that is a great question. I wondered that myself, and sadly to my knowledge, I don't think it is possible to use Legendre's method (the method used in `RcppAlgos::primeCount`) to count primes between two bounds.

I think your best bet is to call `primeCount` first then use `primeSieve` in chunks. That sounds like what you are already doing.

Anywho, I hope that helps.

Best of luck,
Joseph

**Comment by srdeaton** [comment]
Joseph,

 I will definitely mention Kim, could not have done this work with out you
guys. At least in a reasonable time.

Had same thought about Legendre's method, but was not sure it could be done
another way; might be next thing I look into.

thanks,
Shaun
